build: add tooling for changelog generation

Planeshifter opened the pull request:
## Description

> What is the purpose of this pull request?

This pull request:

-   adds tooling for generating changelog files for project packages.

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

Planeshifter committed:
build: changelog generation

Planeshifter committed:
build: add new dev dependencies

Planeshifter committed:
build: add section tags

Planeshifter committed:
build: add version suffix to ids and use kebab case consistently

Planeshifter committed:
build: add parameter to generate changelog for new release

Planeshifter committed:
build: generate changelog in publish script

Planeshifter committed:
build: fix generated ids in changelog

Planeshifter committed:
build: rewrite changelog in standalone repo publish action

Planeshifter committed:
build: refactor to also include breaking changes and closed issues at top-level for namespace packages

Planeshifter committed:
build: use headers for packages in namespace changelogs

Planeshifter committed:
build: add section tag helpers and minor clean-up

Planeshifter committed:
build: extract co-authors from footer

kgryte commented on lib/node_modules/@stdlib/_tools/changelog/generate/README.md:
```diff
@@ -0,0 +1,109 @@
+var changelog = generate( 'assert/contains', 'patch' );
+// returns '...'
+
+changelog = generate( 'assert/contains', 'minor' );
```
```suggestion
changelog = generate( 'assert/contains', 'minor' );
// returns '...'
```

kgryte committed:
Apply suggestions from code review

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/_tools/changelog/parse-commits/package.json:
```diff
@@ -0,0 +1,62 @@
+      "url": "https://github.com/stdlib-js/stdlib/graphs/contributors"
+    }
+  ],
+  "bin": {},
```
```suggestion
```

kgryte commented on lib/node_modules/@stdlib/_tools/changelog/generate/package.json:
```diff
@@ -0,0 +1,63 @@
+      "url": "https://github.com/stdlib-js/stdlib/graphs/contributors"
+    }
+  ],
+  "bin": {},
```
```suggestion
```

kgryte committed:
Apply suggestions from code review

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/_tools/changelog/generate/README.md:
````diff
@@ -0,0 +1,110 @@
+Generates a Markdown formatted changelog for a specified package.
+
+```javascript
+var changelog = generate( 'assert/contains' );
````
@Planeshifter I believe you omitted the `@stdlib/` prefix based on how publishing works, correct? However, omitting the prefix is not common convention in other tools (e.g., https://github.com/stdlib-js/stdlib/tree/develop/lib/node_modules/%40stdlib/_tools/pkgs/name2standalone).

Additionally, to be clear, this tool seems, based on the docs alone, to be oriented toward `stdlib` packages, correct? The omission of the `@stdlib/` prefix implies this, but visually a dev may not recognize this.

My personal preference would be to require the `@stdlib/` prefix, which should be a minor adjustment in the publish pipeline.

kgryte replied:
In general, I'd argue for writing this tooling as much as possible on the premise that `stdlib` is a potential consumer, but not the only consumer. In general, I'd like to avoid implicit coupling between this tool, for example, and some other tool/script of which a casual reader will know nothing about.

Planeshifter replied:
Refactored to require the `@stdlib/` prefix.

kgryte commented on lib/node_modules/@stdlib/_tools/changelog/parse-commits/README.md:
````diff
@@ -0,0 +1,94 @@
+Parses package commit messages into conventional changelog objects.
+
+```javascript
+var commits = parseCommits( 'assert/contains' );
````
Same comment https://github.com/stdlib-js/stdlib/pull/1289/files#r1490573190.

Planeshifter replied:
Addressed.

kgryte commented on lib/node_modules/@stdlib/_tools/changelog/parse-commits/README.md:
```diff
@@ -0,0 +1,94 @@
+// returns [...]
+
+// Commits for entire project:
+commits = parseCommits( '' );
```
The use of an empty string is odd and not obvious outside of this README. Similar to https://github.com/stdlib-js/stdlib/pull/1289/files#r1490573190, this seems to assume `stdlib` packages. If so, I'd rather omit the argument entirely, rather than use an empty string.

kgryte replied:
As mentioned in https://github.com/stdlib-js/stdlib/pull/1289/files#r1490628751, I'd decouple this package from the notion of a "package". It should just take a directory path, perhaps even defaulting to the current working directory if a path string is not provided.

Planeshifter replied:
Package has been refactored to not be tied to packages anymore.

kgryte commented on lib/node_modules/@stdlib/_tools/changelog/parse-commits/examples/index.js:
```diff
@@ -0,0 +1,45 @@
+// => [...]
+
+// Commits for entire project:
+commits = parseCommits( '' );
```
This would need to be updated if we decide to move away from `''`.

kgryte commented on lib/node_modules/@stdlib/_tools/changelog/parse-commits/lib/main.js:
```diff
@@ -0,0 +1,87 @@
+				merged.notes[ j ].hash = commits[ i ].hash;
+			}
+			out.push( merged );
+		}
+		catch ( err ) {
```
```suggestion
		} catch ( err ) {
```

kgryte commented on lib/node_modules/@stdlib/_tools/changelog/parse-commits/lib/main.js:
```diff
@@ -0,0 +1,87 @@
+		}
+		catch ( err ) {
+			debug( 'Encountered an error when parsing a commit message: %s. Message: %s', err.message, commits[ i ] );
+			continue;
```
```suggestion
```

kgryte commented on lib/node_modules/@stdlib/_tools/changelog/parse-commits/lib/main.js:
```diff
@@ -0,0 +1,87 @@
+	}
+	commits = extractCommits( pkg );
+	out = [];
+	i = 0;
+
```
```suggestion
```

kgryte committed:
Apply suggestions from code review

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/_tools/changelog/parse-commits/lib/main.js:
```diff
@@ -0,0 +1,83 @@
+	out = [];
+	for ( i = 0; i < commits.length; i++ ) {
+		try {
+			msg = commits[ i ].message;
+			ast = parser( msg );
+			merged = merge( commits[ i ], toConventionalChangelog( ast ) );
+			for ( j = 0; j < merged.notes.length; j++ ) {
+				merged.notes[ j ].hash = commits[ i ].hash;
+			}
+			out.push( merged );
```
You have multiple lines in this `try` block. It would be good to extract out only those lines for which we believe are allowed to throw. Otherwise, you are swallowing and conflating syntax errors with runtime errors.

Planeshifter replied:
Addressed.

kgryte commented on lib/node_modules/@stdlib/_tools/changelog/parse-commits/lib/conventional_changelog.js:
```diff
@@ -0,0 +1,276 @@
+		* @private
+		* @param {Object} node - AST node
+		*/
+		function extractBody( node ) {
```
This function can be moved to parent scope, as it does not use any variables from the scope one-level up.

Planeshifter replied:
Done.

kgryte commented on lib/node_modules/@stdlib/_tools/changelog/parse-commits/lib/conventional_changelog.js:
```diff
@@ -0,0 +1,276 @@
+		* @private
+		* @param {Object} node - AST node
+		*/
+		function visitSummary( node ) {
```
Same comment. This function should be moved on-level up. Otherwise, for every `onSummary` call, you are creating a new function, which increases memory usage and creates additional GC pressure.

Planeshifter replied:
Done.

kgryte commented on lib/node_modules/@stdlib/_tools/changelog/parse-commits/lib/conventional_changelog.js:
```diff
@@ -0,0 +1,276 @@
+	* @param {Object} node - AST node
+	*/
+	function processFooter( node ) {
+		var hasRefSepartor = false;
```
Separator?

kgryte commented on lib/node_modules/@stdlib/_tools/changelog/parse-commits/lib/conventional_changelog.js:
```diff
@@ -0,0 +1,276 @@
+		var reference = {};
+		var hasPRURL = false;
+		visit( node, [ 'type', 'separator', 'text' ], processFooterNode );
+		if ( hasRefSepartor && reference.ref.match( RE_DIGITS ) ) {
```
Why use `match` here? Shouldn't you use `RE_DIGITS.test()`? You're not doing anything with the `match` return value which is wasteful from a micro-perf POV.

Planeshifter replied:
Changed!

kgryte commented on lib/node_modules/@stdlib/_tools/changelog/parse-commits/lib/conventional_changelog.js:
```diff
@@ -0,0 +1,276 @@
+
+// VARIABLES //
+
+var STDLIB_ISSUE_URL = 'https://github.com/stdlib-js/stdlib/issues/';
```
Rather than hardcode these URLs, it would be better to pass in as options to the main export (e.g., `opts.issueURL` and `opts.prURL` or `opts.issue_url` and `opts.pr_url`). This would help make this package more general.

Planeshifter replied:
This has meanwhile been implemented.

kgryte commented on lib/node_modules/@stdlib/_tools/changelog/parse-commits/lib/commits.js:
```diff
@@ -0,0 +1,111 @@
+		'maxBuffer': 1024*1024*60, // 60 MB
+		'stdio': ['pipe', 'pipe', 'ignore'] // stdin, stdout, stderr
+	};
+	cmd = 'if [ -d "lib/node_modules/@stdlib/'+pkg+'" ]; then\n';
+	cmd = '  git log --name-only --no-merges --pretty=format:"%H|%ad|%aN <%aE>|%B|"';
+	cmd += ' ';
+	cmd += 'lib/node_modules/@stdlib/'+pkg;
+
+	// Insert a separator between commits:
+	cmd += ' | ';
+	cmd += 'awk \'/^$/{p=1;next} /^[0-9a-f]{40}\\|/{if (p==1) print "'+GIT_COMMIT_SEP+'"; p=0} {print}\'';
+	cmd += 'fi\n';
```
Move this to a separate script which takes a single argument and load the file contents using `@stdlib/fs/read-file`. We **really** need to move away from writing scripts this way. We explicitly use shellcheck to help catch lint errors and embedding content this way makes debugging more onerous than it needs to be.

kgryte commented on lib/node_modules/@stdlib/_tools/changelog/parse-commits/lib/commits.js:
```diff
@@ -0,0 +1,111 @@
+	var i;
+
+	opts = {
+		'cwd': rootDir(),
```
This should be configurable. Similar to other tooling, make the directory from which to search for packages an input option.

Planeshifter replied:
Addressed.

kgryte commented on lib/node_modules/@stdlib/_tools/changelog/parse-commits/lib/commits.js:
```diff
@@ -0,0 +1,111 @@
+		'maxBuffer': 1024*1024*60, // 60 MB
+		'stdio': ['pipe', 'pipe', 'ignore'] // stdin, stdout, stderr
+	};
+	cmd = 'if [ -d "lib/node_modules/@stdlib/'+pkg+'" ]; then\n';
+	cmd = '  git log --name-only --no-merges --pretty=format:"%H|%ad|%aN <%aE>|%B|"';
```
```suggestion
	cmd = 'if [ -d "lib/node_modules/@stdlib/'+pkg+'" ]; then\n';
	cmd = '  git log --name-only --no-merges --pretty=format:"%H|%ad|%aN <%aE>|%B|"';
```

For example, what is happening in these lines? You're assigning to `cmd` twice. Such mistakes would be easier to catch if this was written in an external shell script.

kgryte commented on lib/node_modules/@stdlib/_tools/changelog/parse-commits/lib/commits.js:
```diff
@@ -0,0 +1,111 @@
+	cmd = 'if [ -d "lib/node_modules/@stdlib/'+pkg+'" ]; then\n';
+	cmd = '  git log --name-only --no-merges --pretty=format:"%H|%ad|%aN <%aE>|%B|"';
+	cmd += ' ';
+	cmd += 'lib/node_modules/@stdlib/'+pkg;
```
If `dir` is configurable, one does not need to specify the package name at all. In fact, the main export can just be a directory path. One does not need the concept of a "package" to use this package. Give this package a directory, it runs `git log` and returns some processed commits.

In short, I'd decouple this tool from the notion of a package. It's a way of parsing commit logs for a specified directory.

kgryte commented on lib/node_modules/@stdlib/_tools/changelog/generate/lib/release_summary.js:
```diff
@@ -0,0 +1,113 @@
+	* @param {Object} commit - commit object
+	* @returns {(string|null)} group key
+	*/
+	function groupByType( commit ) {
```
This function does not depend on variables in the current scope. This should be moved to parent scope.

Planeshifter replied:
Done.

kgryte commented on lib/node_modules/@stdlib/_tools/changelog/generate/lib/npm_releases.js:
```diff
@@ -0,0 +1,73 @@
+		out = shell( command, opts ).toString();
+		out = JSON.parse( out );
+		out = omit( out, OMITTED_KEYS );
+	}
+	catch ( err ) {
```
```suggestion
	} catch ( err ) {
```

We should create a lint rule that only allows breaks like this when there is a `// Comment` after the associated clause.

kgryte commented on lib/node_modules/@stdlib/_tools/changelog/generate/lib/npm_releases.js:
```diff
@@ -0,0 +1,73 @@
+		'stdio': ['pipe', 'pipe', 'ignore'] // stdin, stdout, stderr
+	};
+	try {
+		out = shell( command, opts ).toString();
+		out = JSON.parse( out );
+		out = omit( out, OMITTED_KEYS );
```
```suggestion
		out = shell( command, opts ).toString();
		out = JSON.parse( out );
		out = omit( out, OMITTED_KEYS );
```

We should move whatever expressions we do not except to error out of the `try/catch`; otherwise, you risk conflating syntax and operational errors.

kgryte commented on lib/node_modules/@stdlib/_tools/changelog/generate/lib/main.js:
```diff
@@ -0,0 +1,388 @@
+
+/**
+* Generates a Markdown formatted changelog for a specified package.
+*
```
@Planeshifter Looks like we are getting JSDoc doctest lint errors here.

kgryte commented on lib/node_modules/@stdlib/_tools/changelog/generate/lib/main.js:
```diff
@@ -0,0 +1,388 @@
+			if ( !releaseCommits ) {
+				continue;
+			}
+			date = new Date( releases[ i ][ 1 ] ).toISOString().slice( 0, 10 );
```
I believe you should be able to repurpose `formatDate` above, with possibly a small tweak.

Planeshifter replied:
Addressed.

kgryte commented on lib/node_modules/@stdlib/_tools/changelog/generate/lib/main.js:
```diff
@@ -0,0 +1,388 @@
+			if ( !summary ) {
+				continue;
+			}
+			date = new Date( releases[ i ][ 1 ] ).toISOString().slice( 0, 10 );
```
Same comment.

Planeshifter replied:
Addressed.

kgryte commented on lib/node_modules/@stdlib/_tools/changelog/generate/lib/format_contributors.js:
```diff
@@ -0,0 +1,119 @@
+	out += heading( 'Contributors', 3 );
+	if ( contributors.length === 1 ) {
+		out += 'A total of 1 person contributed to this release. Thank you to this contributor:\n\n';
+	}
+	else {
```
```suggestion
	} else {
```

kgryte commented on lib/node_modules/@stdlib/_tools/changelog/generate/lib/format_commit_factory.js:
```diff
@@ -0,0 +1,111 @@
+				out += '(' + commit.scope + ')';
+			}
+			out += ':** ';
+		}
+		else if ( commit.scope ) {
```
```suggestion
		} else if ( commit.scope ) {
```

kgryte commented on lib/node_modules/@stdlib/_tools/changelog/generate/lib/format_commit_factory.js:
```diff
@@ -0,0 +1,111 @@
+		function onMention( mention ) {
+			if ( mention.action === 'PR-URL' ) {
+				out += ' [(#' + trim( mention.ref ) + ')](' + mention.url + ')';
+			}
+			else if ( RE_CO_AUTHORED_BY.test( mention.action ) ) {
```
```suggestion
			} else if ( RE_CO_AUTHORED_BY.test( mention.action ) ) {
```

kgryte commented on lib/node_modules/@stdlib/_tools/changelog/generate/lib/closed_issues.js:
```diff
@@ -0,0 +1,120 @@
+	mentions = collectField( commits, 'mentions' );
+	closedIssues = filter( mentions, isClosingIssue );
+	closedIssues.sort( sortByIssueNumber );
+	if ( !closedIssues || closedIssues.length === 0 ) {
+		return '';
+	}
```
Wouldn't it make sense to shift up this check? And when is `closedIssues` falsy?

Planeshifter replied:
Moved and ensured that `closedIssues` are always an array (this was not the case when grouping commits and not having commits matching the `unreleased` group...)

kgryte commented on lib/node_modules/@stdlib/_tools/changelog/generate/lib/breaking_changes.js:
```diff
@@ -0,0 +1,110 @@
+	* @param {Object} note - note object
+	* @returns {boolean} boolean indicating whether a note is a breaking change
+	*/
+	function isBreakingChange( note ) {
```
Should be moved to parent scope.

Planeshifter replied:
Done.

kgryte commented on lib/node_modules/@stdlib/_tools/changelog/generate/lib/breaking_changes.js:
```diff
@@ -0,0 +1,110 @@
+
+	notes = collectField( commits, 'notes' );
+	breakingChanges = filter( notes, isBreakingChange );
+	if ( !breakingChanges || breakingChanges.length === 0 ) {
```
Same comment as below. When is `breakingChanges` falsy? Isn't it only an empty array?

Planeshifter replied:
Changed.

kgryte commented on lib/node_modules/@stdlib/_tools/changelog/generate/lib/main.js:
```diff
@@ -0,0 +1,388 @@
+	out = '# CHANGELOG\n\n';
+	out += '> Package changelog.\n\n';
+
+	commits = parseCommits( pkg );
```
For this line, `parseCommits` does not need to be "package" aware. One can convert the package name to a directory path.

kgryte commented on lib/node_modules/@stdlib/_tools/changelog/generate/lib/main.js:
```diff
@@ -0,0 +1,388 @@
+	out += '> Package changelog.\n\n';
+
+	commits = parseCommits( pkg );
+	if ( !commits || commits.length === 0 ) {
```
Is `commits` ever `null`ish? Or only an empty array?

Planeshifter replied:
Changed.

kgryte commented on lib/node_modules/@stdlib/_tools/changelog/generate/README.md:
```diff
@@ -0,0 +1,110 @@
+
+# Generate Changelog
+
+> Generate a changelog for a specified package.
```
```suggestion
> Generate a changelog for a specified stdlib package.
```

This package is implemented (and intended to be) very `stdlib` aware. So we should clarify that this is specifically for generating a stdlib package changelog here and elsewhere.

kgryte committed:
Apply suggestions from code review

Signed-off-by: Athan <kgryte@gmail.com>

kgryte reviewed (changes requested):
@Planeshifter Thanks for working on this. At a high level, this is looking good. I've reviewed and left comments. I think it would be good to decouple `parse-commits` from the notion of a "package". And then there are various nits and questions I've left sprinkled throughout.

Planeshifter committed:
build: implement suggestions from code review

Planeshifter committed:
Update lib/node_modules/@stdlib/_tools/changelog/generate/lib/npm_releases.js

Co-authored-by: Athan <kgryte@gmail.com>
Signed-off-by: Philipp Burckhardt <pburckhardt@outlook.com>

Planeshifter committed:
Update lib/node_modules/@stdlib/_tools/changelog/generate/lib/format_contributors.js

Co-authored-by: Athan <kgryte@gmail.com>
Signed-off-by: Philipp Burckhardt <pburckhardt@outlook.com>

Planeshifter committed:
Update lib/node_modules/@stdlib/_tools/changelog/generate/README.md

Co-authored-by: Athan <kgryte@gmail.com>
Signed-off-by: Philipp Burckhardt <pburckhardt@outlook.com>

Planeshifter committed:
build: expect package names with stdlib namespace prefix

Planeshifter committed:
refactor: implement code review feedback

Planeshifter committed:
build: refactor to accept directories instead of packages

Planeshifter committed:
build: fix passed arguments

kgryte commented:
@Planeshifter Once this is ready for another review, LMK in this thread.

kgryte commented:
...or just re-request the review using the UI.

Planeshifter committed:
build: add tool to suggest version change

Planeshifter committed:
build: refactor to allow for automatically determined version bumps

Planeshifter committed:
build: move git log cmd to standalone script

Planeshifter committed:
build: minor doc fixes and clean-up

Planeshifter commented:
@kgryte The PR is now ready for review after addressing all the feedback from the code review.

I have added an additional package, `@stdlib/_tools/changelog/recommend-version-bump`, for suggesting a version bump based on a list of commits, and reworked the `@stdlib/_tools/changelog/generate` return value and publish script to support a new `auto` option that when enabled will automatically determine the release type in the publishing process (or skip a new release if no new release is warranted.)

Another review would be appreciated!

kgryte commented:
Thanks, @Planeshifter! I'll try and get another review in before tomorrow.

kgryte commented on lib/node_modules/@stdlib/_tools/changelog/parse-commits/test/test.validate.js:
```diff
@@ -0,0 +1,179 @@
+	t.end();
+});
+
+tape( 'if provided a `issueURL` option which is not a `string`, the function returns a type error', function test( t ) {
```
```suggestion
tape( 'if provided an `issueURL` option which is not a `string`, the function returns a type error', function test( t ) {
```

kgryte committed:
Apply suggestions from code review

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/_tools/changelog/generate/lib/npm_releases.js:
```diff
@@ -0,0 +1,72 @@
+	command = 'npm view ' + pkg + ' time --json';
+	opts = {
+		'cwd': rootDir(),
+		'stdio': ['pipe', 'pipe', 'ignore'] // stdin, stdout, stderr
```
```suggestion
		'stdio': [ 'pipe', 'pipe', 'ignore' ] // stdin, stdout, stderr
```

kgryte committed:
Apply suggestions from code review

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/_tools/changelog/generate/lib/format_contributors.js:
```diff
@@ -0,0 +1,118 @@
+				!contains( out, mention.ref ) &&
+				!contains( EXCLUDED_CONTRIBUTORS, mention.ref )
+			) {
+				out.push( mentions[ j ].ref );
```
```suggestion
				out.push( mention.ref );
```

kgryte committed:
Apply suggestions from code review

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/_tools/changelog/parse-commits/lib/commits.js:
```diff
@@ -0,0 +1,115 @@
+
+	opts = {
+		'maxBuffer': 1024*1024*60, // 60 MB
+		'stdio': ['pipe', 'pipe', 'ignore'], // stdin, stdout, stderr
```
```suggestion
		'stdio': [ 'pipe', 'pipe', 'ignore' ], // stdin, stdout, stderr
```

kgryte commented on lib/node_modules/@stdlib/_tools/changelog/parse-commits/lib/commits.js:
```diff
@@ -0,0 +1,115 @@
+	};
+	cmd = join( __dirname, '..', 'scripts', 'commits.sh' );
+	commits = [];
+	try {
+		args = [ resolve( dir ) ];
+		out = spawn( cmd, args, opts );
+		out = trim( out.stdout.toString() );
+	} catch ( err ) {
+		debug( 'Encountered an error when attempting to extract commits: %s', err.message );
+		return commits;
+	}
```
```suggestion
	args = [ resolve( dir ) ];
	try {
		out = spawn( cmd, args, opts );
	} catch ( err ) {
		debug( 'Encountered an error when attempting to extract commits: %s', err.message );
		return commits;
	}
	out = trim( out.stdout.toString() );
```

Only wrap logic in a try-catch which can reasonably be expected to potentially error. Otherwise, you can mask syntax errors, etc.

kgryte committed:
Apply suggestions from code review

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/_tools/changelog/parse-commits/lib/main.js:
```diff
@@ -0,0 +1,104 @@
+		try {
+			ast = parser( msg );
+		} catch ( err ) {
+			debug( 'Encountered an error when parsing a commit message: %s. Message: %s', err.message, commits[ i ] );
```
```suggestion
			debug( 'Encountered an error when parsing a commit message: %s. Error: %s', msg, err.message );
```

I believe you are wanting to log the failed commit message here, not the commit object.

kgryte committed:
Apply suggestions from code review

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/_tools/changelog/parse-commits/test/test.js:
```diff
@@ -0,0 +1,116 @@
+		'dir': 'beepboop'
+	};
+	var out = parseCommits( opts );
+	t.strictEqual( isArray( out ), true, 'returns an array' );
```
```suggestion
	t.strictEqual( isArray( out ), true, 'returns expected value' );
```

kgryte commented on lib/node_modules/@stdlib/_tools/changelog/parse-commits/test/test.js:
```diff
@@ -0,0 +1,116 @@
+	};
+	var out = parseCommits( opts );
+	t.strictEqual( isArray( out ), true, 'returns an array' );
+	t.strictEqual( out.length, 0, 'has length 0' );
```
```suggestion
	t.strictEqual( out.length, 0, 'returns expected value' );
```

kgryte commented on lib/node_modules/@stdlib/_tools/changelog/parse-commits/test/test.js:
```diff
@@ -0,0 +1,116 @@
+		'dir': join( rootDir(), 'lib', 'node_modules', '@stdlib', 'utils', 'curry' )
+	};
+	commits = parseCommits( opts );
+	t.strictEqual( isArray( commits ), true, 'returns an array' );
+	t.strictEqual( commits.length > 0, true, 'has length greater than 0' );
+	for ( i = 0; i < commits.length; i++ ) {
+		commit = commits[ i ];
+		t.strictEqual( isPlainObject( commit ), true, 'element is an object' );
+		t.strictEqual( hasOwnProp( commit, 'hash' ), true, 'has property' );
+		t.strictEqual( hasOwnProp( commit, 'date' ), true, 'has property' );
+		t.strictEqual( hasOwnProp( commit, 'author' ), true, 'has property' );
+		t.strictEqual( hasOwnProp( commit, 'message' ), true, 'has property' );
+		t.strictEqual( hasOwnProp( commit, 'type' ), true, 'has property' );
+		t.strictEqual( hasOwnProp( commit, 'scope' ), true, 'has property' );
+		t.strictEqual( hasOwnProp( commit, 'subject' ), true, 'has property' );
+		t.strictEqual( hasOwnProp( commit, 'body' ), true, 'has property' );
+		t.strictEqual( hasOwnProp( commit, 'footer' ), true, 'has property' );
+		t.strictEqual( hasOwnProp( commit, 'notes' ), true, 'has property' );
+		t.strictEqual( hasOwnProp( commit, 'references' ), true, 'has property' );
```
```suggestion
	t.strictEqual( isArray( commits ), true, 'returns an array' );
	t.strictEqual( commits.length > 0, true, 'has length greater than 0' );
	for ( i = 0; i < commits.length; i++ ) {
		commit = commits[ i ];
		t.strictEqual( isPlainObject( commit ), true, 'returns expected value' );
		t.strictEqual( hasOwnProp( commit, 'hash' ), true, 'returns expected value' );
		t.strictEqual( hasOwnProp( commit, 'date' ), true, 'returns expected value' );
		t.strictEqual( hasOwnProp( commit, 'author' ), true, 'returns expected value' );
		t.strictEqual( hasOwnProp( commit, 'message' ), true, 'returns expected value' );
		t.strictEqual( hasOwnProp( commit, 'type' ), true, 'returns expected value' );
		t.strictEqual( hasOwnProp( commit, 'scope' ), true, 'returns expected value' );
		t.strictEqual( hasOwnProp( commit, 'subject' ), true, 'returns expected value' );
		t.strictEqual( hasOwnProp( commit, 'body' ), true, 'returns expected value' );
		t.strictEqual( hasOwnProp( commit, 'footer' ), true, 'returns expected value' );
		t.strictEqual( hasOwnProp( commit, 'notes' ), true, 'returns expected value' );
		t.strictEqual( hasOwnProp( commit, 'references' ), true, 'returns expected value' );
```

kgryte committed:
Apply suggestions from code review

Signed-off-by: Athan <kgryte@gmail.com>

kgryte committed:
Apply suggestions from code review

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/_tools/changelog/parse-commits/test/test.js:
```diff
@@ -0,0 +1,116 @@
+tape( 'the function throws an error if provided an invalid option', function test( t ) {
+	t.throws( badValue, TypeError, 'throws error' );
+	t.end();
+	function badValue() {
```
```suggestion

	function badValue() {
```

kgryte commented on lib/node_modules/@stdlib/_tools/changelog/parse-commits/test/test.js:
```diff
@@ -0,0 +1,116 @@
+		'dir': join( rootDir(), 'lib', 'node_modules', '@stdlib', 'utils', 'curry' )
+	};
+	commits = parseCommits( opts );
+	t.strictEqual( isArray( commits ), true, 'returns an array' );
```
```suggestion
	t.strictEqual( isArray( commits ), true, 'returns expected value' );
```

kgryte commented on lib/node_modules/@stdlib/_tools/changelog/parse-commits/test/test.js:
```diff
@@ -0,0 +1,116 @@
+	};
+	commits = parseCommits( opts );
+	t.strictEqual( isArray( commits ), true, 'returns an array' );
+	t.strictEqual( commits.length > 0, true, 'has length greater than 0' );
```
```suggestion
	t.strictEqual( commits.length > 0, true, 'returns expected value' );
```

kgryte committed:
Apply suggestions from code review

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/_tools/changelog/recommend-version-bump/README.md:
````diff
@@ -0,0 +1,118 @@
+// returns 'minor'
+```
+
+The function returns a `string` indicating the recommended version bump or `null` if no version bump is recommended:
````
```suggestion
The function returns a string indicating the recommended version bump or `null` if no version bump is recommended:
```

kgryte committed:
Apply suggestions from code review

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/_tools/changelog/recommend-version-bump/lib/main.js:
```diff
@@ -0,0 +1,71 @@
+/**
+* Recommends a version bump based on a list of commits.
+*
+* @param {Array} commits - conventional changelog formatted commit message objects
```
```suggestion
* @param {Array<Object>} commits - conventional changelog formatted commit message objects
```

kgryte commented on lib/node_modules/@stdlib/_tools/changelog/recommend-version-bump/lib/main.js:
```diff
@@ -0,0 +1,71 @@
+* @returns {(string|null)} - `major`, `minor`, `patch`, or null
+*/
+function recommendVersionBump( commits ) {
+	var hasBreakingChange = false;
+	var hasFeature = false;
+	var hasFix = false;
```
```suggestion
	var hasBreakingChange;
	var hasFeature;
	var hasFix;
```

kgryte committed:
Apply suggestions from code review

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/_tools/changelog/recommend-version-bump/lib/main.js:
```diff
@@ -0,0 +1,71 @@
+		commit = commits[ i ];
+		for ( j = 0; j < commit.notes.length; j++ ) {
+			if ( commit.notes[ j ].title === 'BREAKING CHANGE' ) {
+				hasBreakingChange = true;
+				break;
```
```suggestion
				return 'major';
```

kgryte commented on lib/node_modules/@stdlib/_tools/changelog/recommend-version-bump/lib/main.js:
```diff
@@ -0,0 +1,71 @@
+		} else if ( commit.type === 'fix' ) {
+			hasFix = true;
+		}
+
+		// If a breaking change is found, no need to check further commits:
+		if ( hasBreakingChange ) {
+			break;
+		}
+	}
+	if ( hasBreakingChange ) {
+		return 'major';
+	}
```
```suggestion
	}
```

kgryte commented on lib/node_modules/@stdlib/_tools/changelog/recommend-version-bump/lib/main.js:
```diff
@@ -0,0 +1,71 @@
+* @returns {(string|null)} - `major`, `minor`, `patch`, or null
+*/
+function recommendVersionBump( commits ) {
+	var hasBreakingChange;
```
```suggestion
```

kgryte committed:
Apply suggestions from code review

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/_tools/changelog/recommend-version-bump/test/test.js:
```diff
@@ -0,0 +1,94 @@
+
+tape( 'the function returns `null` for an empty array of commits', function test( t ) {
+	var result = recommendVersionBump( [] );
+	t.equal( result, null, 'returns null for empty commit array' );
```
```suggestion
	t.equal( result, null, 'returns expected value' );
```

@Planeshifter Avoid accepting copilot's suggestion for test message assertions, please. This just creates unnecessary work.

kgryte commented on lib/node_modules/@stdlib/_tools/changelog/recommend-version-bump/test/test.js:
```diff
@@ -0,0 +1,94 @@
+			'notes': []
+		}
+	];
+	var result = recommendVersionBump(commits);
+	t.equal( result, 'patch', 'returns `patch` for fix commits' );
```
```suggestion
	var result = recommendVersionBump( commits );
	t.equal( result, 'patch', 'returns expected value' );
```

kgryte commented on lib/node_modules/@stdlib/_tools/changelog/recommend-version-bump/test/test.js:
```diff
@@ -0,0 +1,94 @@
+			'notes': []
+		}
+	];
+	var result = recommendVersionBump(commits);
+	t.equal( result, 'minor', 'returns `minor` for feature commits' );
```
```suggestion
	var result = recommendVersionBump( commits );
	t.equal( result, 'minor', 'returns expected value' );
```

kgryte commented on lib/node_modules/@stdlib/_tools/changelog/recommend-version-bump/test/test.js:
```diff
@@ -0,0 +1,94 @@
+			}]
+		}
+	];
+	var result = recommendVersionBump(commits);
+	t.equal( result, 'major', 'returns `major` for breaking change commits' );
```
```suggestion
	var result = recommendVersionBump( commits );
	t.equal( result, 'major', 'returns expected value' );
```

kgryte committed:
Apply suggestions from code review

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/_tools/changelog/generate/test/test.js:
```diff
@@ -0,0 +1,157 @@
+	out = generate( '@stdlib/stats/base/dists' );
+	t.strictEqual( isObject( out ), true, 'returns an object' );
+	t.strictEqual( typeof out.content, 'string', 'has `content` property' );
+	t.strictEqual( out.releaseType, null, 'has `releaseType` property' );
+
+	out = generate( '@stdlib/proxy' );
+	t.strictEqual( isObject( out ), true, 'returns an object' );
+	t.strictEqual( typeof out.content, 'string', 'has `content` property' );
+	t.strictEqual( out.releaseType, null, 'has `releaseType` property' );
+
+	t.end();
+});
+
+tape( 'the function generates a changelog for a new release', function test( t ) {
+	var out = generate( '@stdlib/utils/noop', 'patch' );
+	t.strictEqual( isObject( out ), true, 'returns an object' );
+	t.strictEqual( typeof out.content, 'string', 'has `content` property' );
+	t.strictEqual( out.releaseType, 'patch', 'has `releaseType` property' );
+
+	out = generate( '@stdlib/utils/noop', 'minor' );
+	t.strictEqual( isObject( out ), true, 'returns an object' );
+	t.strictEqual( typeof out.content, 'string', 'has `content` property' );
+	t.strictEqual( out.releaseType, 'minor', 'has `releaseType` property' );
+
+	out = generate( '@stdlib/utils/noop', 'major' );
+	t.strictEqual( isObject( out ), true, 'returns an object' );
+	t.strictEqual( typeof out.content, 'string', 'has `content` property' );
+	t.strictEqual( out.releaseType, 'major', 'has `releaseType` property' );
+
+	t.end();
+});
+
+tape( 'the function generates a changelog for a new release (auto)', function test( t ) {
+	var out = generate( '@stdlib/utils/noop', 'auto' );
+	t.strictEqual( isObject( out ), true, 'returns an object' );
+	t.strictEqual( typeof out.content, 'string', 'has `content` property' );
+
+	out = generate( '@stdlib/array/base/slice', 'auto' );
+	t.strictEqual( isObject( out ), true, 'returns an object' );
+	t.strictEqual( typeof out.content, 'string', 'has `content` property' );
```
```suggestion
	t.strictEqual( isObject( out ), true, 'returns expected value' );
	t.strictEqual( typeof out.content, 'string', 'returns expected value' );
	t.strictEqual( out.releaseType, null, 'returns expected value' );

	out = generate( '@stdlib/stats/base/dists' );
	t.strictEqual( isObject( out ), true, 'returns expected value' );
	t.strictEqual( typeof out.content, 'string', 'returns expected value' );
	t.strictEqual( out.releaseType, null, 'returns expected value' );

	out = generate( '@stdlib/proxy' );
	t.strictEqual( isObject( out ), true, 'returns expected value' );
	t.strictEqual( typeof out.content, 'string', 'returns expected value' );
	t.strictEqual( out.releaseType, null, 'returns expected value' );

	t.end();
});

tape( 'the function generates a changelog for a new release', function test( t ) {
	var out = generate( '@stdlib/utils/noop', 'patch' );
	t.strictEqual( isObject( out ), true, 'returns expected value' );
	t.strictEqual( typeof out.content, 'string', 'returns expected value' );
	t.strictEqual( out.releaseType, 'patch', 'returns expected value' );

	out = generate( '@stdlib/utils/noop', 'minor' );
	t.strictEqual( isObject( out ), true, 'returns expected value' );
	t.strictEqual( typeof out.content, 'string', 'returns expected value' );
	t.strictEqual( out.releaseType, 'minor', 'returns expected value' );

	out = generate( '@stdlib/utils/noop', 'major' );
	t.strictEqual( isObject( out ), true, 'returns expected value' );
	t.strictEqual( typeof out.content, 'string', 'returns expected value' );
	t.strictEqual( out.releaseType, 'major', 'returns expected value' );

	t.end();
});

tape( 'the function generates a changelog for a new release (auto)', function test( t ) {
	var out = generate( '@stdlib/utils/noop', 'auto' );
	t.strictEqual( isObject( out ), true, 'returns expected value' );
	t.strictEqual( typeof out.content, 'string', 'returns expected value' );

	out = generate( '@stdlib/array/base/slice', 'auto' );
	t.strictEqual( isObject( out ), true, 'returns expected value' );
	t.strictEqual( typeof out.content, 'string', 'returns expected value' );
```

kgryte committed:
Apply suggestions from code review

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/_tools/changelog/parse-commits/lib/conventional_changelog.js:
```diff
@@ -0,0 +1,277 @@
+			case 'type':
+				if ( node.value === 'PR-URL' ) {
+					hasPRURL = true;
+				}
+				else if ( RE_CLOSES.test( node.value ) ) {
+					closesIssue = true;
+				}
+				else if ( RE_CO_AUTHORED_BY.test( node.value ) ) {
```
```suggestion
				} else if ( RE_CLOSES.test( node.value ) ) {
					closesIssue = true;
				} else if ( RE_CO_AUTHORED_BY.test( node.value ) ) {
```

kgryte commented on lib/node_modules/@stdlib/_tools/changelog/parse-commits/lib/conventional_changelog.js:
```diff
@@ -0,0 +1,277 @@
+		var closesIssue = false;
+		var reference = {};
+		var hasPRURL = false;
+		visit( node, [ 'type', 'separator', 'text' ], processFooterNode );
```
```suggestion

		visit( node, [ 'type', 'separator', 'text' ], processFooterNode );
```

kgryte commented on lib/node_modules/@stdlib/_tools/changelog/parse-commits/lib/conventional_changelog.js:
```diff
@@ -0,0 +1,277 @@
+				}
+				break;
+			case 'text':
+				if ( node.value.includes( options.issueURL ) ) {
+					reference.url = node.value;
+					reference.ref = node.value.split( '/' ).pop();
+				}
+				else if ( node.value.includes( options.prURL ) ) {
+					hasPRURL = true;
+					reference.url = node.value;
+					reference.ref = node.value.split('/').pop();
+				}
+				else if ( node.value.charAt( 0 ) === '#' ) {
```
```suggestion
				if ( node.value.includes( options.issueURL ) ) {
					reference.url = node.value;
					reference.ref = node.value.split( '/' ).pop();
				} else if ( node.value.includes( options.prURL ) ) {
					hasPRURL = true;
					reference.url = node.value;
					reference.ref = node.value.split('/').pop();
				} else if ( node.value.charAt( 0 ) === '#' ) {
```

Use consistent line breaking. There's no rhyme or reason for why certain linebreaks are present and missing elsewhere.

kgryte committed:
Apply suggestions from code review

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/_tools/changelog/parse-commits/lib/conventional_changelog.js:
```diff
@@ -0,0 +1,274 @@
+				}
+				break;
+			case 'text':
+				if ( node.value.includes( options.issueURL ) ) {
+					reference.url = node.value;
+					reference.ref = node.value.split( '/' ).pop();
+				} else if ( node.value.includes( options.prURL ) ) {
```
```suggestion
				if ( node.value.includes( options.issueURL ) ) {
					reference.url = node.value;
					reference.ref = node.value.split( '/' ).pop();
				} else if ( node.value.includes( options.prURL ) ) {
```

You're using `#includes`. Switch to ES5 equivalent, such as `#indexOf` or a stdlib package equivalent.

Planeshifter replied:
@kgryte Replaced. Ready for final approval now.

kgryte commented on lib/node_modules/@stdlib/_tools/changelog/generate/lib/main.js:
```diff
@@ -0,0 +1,403 @@
+*/
+function isNotInternalTooling( commit ) {
+	var files;
+	var bool;
+	var i;
+
+	files = commit.files;
+	bool = false;
+	for ( i = 0; i < files.length; i++ ) {
+		if ( files[ i ].indexOf( 'lib/node_modules/@stdlib/_tools/' ) === -1 ) {
+			bool = true;
+			break;
+		}
+	}
+	return bool;
```
```suggestion
	var i;

	files = commit.files;
	for ( i = 0; i < files.length; i++ ) {
		if ( files[ i ].indexOf( 'lib/node_modules/@stdlib/_tools/' ) === -1 ) {
			return true;
		}
	}
	return false;
```

kgryte committed:
Apply suggestions from code review

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/_tools/changelog/generate/lib/main.js:
```diff
@@ -0,0 +1,400 @@
+
+	// Group commits by release:
+	commits = groupBy( commits, indicator );
+	if (!commits.unreleased) {
```
```suggestion
	if ( !commits.unreleased ) {
```

kgryte commented on lib/node_modules/@stdlib/_tools/changelog/generate/lib/main.js:
```diff
@@ -0,0 +1,400 @@
+	* @returns {string} release indicator
+	*/
+	function indicator( commit ) {
+		var version;
+		var date;
+		var i;
+
+		// Walk the releases in reverse chronological order:
+		for ( i = releases.length-1; i >= 0; i-- ) {
+			version = releases[ i ][ 0 ];
+			date = releases[ i ][ 1 ];
+			if ( new Date( commit.date ) <= new Date( date ) ) {
+				return version;
```
```suggestion
		var date;
		var i;
		
		date = new Date( commit.date );

		// Walk the releases in reverse chronological order:
		for ( i = releases.length-1; i >= 0; i-- ) {
			if ( date <= new Date( releases[ i ][ 1 ] ) ) {
				return releases[ i ][ 0 ]; // version
```

kgryte committed:
Apply suggestions from code review

Signed-off-by: Athan <kgryte@gmail.com>

kgryte reviewed (changes requested):
@Planeshifter apart from the ES6 which slipped in, this PR should be ready for merge. Once the final comment is resolved, I'll approve and merge.

kgryte commented on lib/node_modules/@stdlib/_tools/changelog/generate/lib/main.js:
```diff
@@ -0,0 +1,399 @@
+	function indicator( commit ) {
+		var date;
+		var i;
+		
```
```suggestion

```

kgryte committed:
style: remove trailing spaces

Signed-off-by: Athan <kgryte@gmail.com>

Planeshifter committed:
fix: replace includes with indexOf checks

kgryte reviewed (approved):
LGTM. Let's go ahead and merge and address any subsequent changes in follow-up PRs...
